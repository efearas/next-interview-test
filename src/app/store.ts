import { configureStore, ThunkAction, Action } from '@reduxjs/toolkit';
import searchResultsReducer from '../features/searchResults/searchResultsSlice';


export const store = configureStore({
  reducer: {    
    searchResults: searchResultsReducer,
  },
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
