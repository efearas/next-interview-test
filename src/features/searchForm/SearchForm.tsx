import React, { useState, useEffect, useRef } from "react";
import { makeStyles } from "@material-ui/core/styles";
import InputLabel from "@material-ui/core/InputLabel";
import MenuItem from "@material-ui/core/MenuItem";
import FormControl from "@material-ui/core/FormControl";
import Select from "@material-ui/core/Select";
import TextField from "@material-ui/core/TextField";
import Button from "@material-ui/core/Button";
import { useAppDispatch, useAppSelector } from "../../app/hooks";
import {
  getSearchResults,
  selectSearchResults,
  EnumSearchStatus,
  changeSearchText,
} from "../searchResults/searchResultsSlice";

export enum EnumSearchType {
  artist = 0,
  album = 1,
  song = 2,
}

const searchTypeArr: EnumSearchType[] = [
  EnumSearchType.artist,
  EnumSearchType.album,
  EnumSearchType.song,
];

export interface ISearchParams {
  entity: string;
  term: string;
  attribute: string;
  offset: number;
}

const useStyles = makeStyles((theme) => ({
  formControl: {
    width: "100%",
    "& > div,button": {
      marginTop: "20px",
      minWidth: "200px",
    },
    "@media(min-width:768px)": {
      display: "flex",
      flexDirection: "row",
      width: "100%",
      maxWidth: "800px",
      justifyContent: "space-around",
      "& > div,button": {
        marginLeft: "20px",
      },
    },
  },
  wrapper: {
    padding: "10px",
  },
}));

const SearchForm = () => {
  const classes = useStyles();
  const dispatch = useAppDispatch();
  const searchResultsState = useAppSelector(selectSearchResults);
  const gettingData = useRef(false);
  const offset = useRef(0);
  const [searchType, setSearchType] = useState<EnumSearchType>(
    EnumSearchType.artist
  );

  useEffect(() => {
    if (searchResultsState.searchStatus === EnumSearchStatus.SearchCompleted) {
      gettingData.current = false;
    }
  }, [searchResultsState.searchStatus]);

  const handleChange = (event: React.ChangeEvent<{ value: unknown }>) => {
    setSearchType(event.target.value as EnumSearchType);
  };

  const search = (isNewSearch: boolean = true) => {
    dispatch(getSearchResults(isNewSearch));
  };

  return (
    <div className={classes.wrapper}>
      <FormControl className={classes.formControl}>
        <InputLabel id="select-search-type">Select search type</InputLabel>
        <Select
          labelId="select-search-type"
          id="select-search-type"
          value={searchResultsState.searchType}
          onChange={handleChange}
        >
          {searchTypeArr.map((value) => (
            <MenuItem key={value} value={value}>
              {EnumSearchType[value]}
            </MenuItem>
          ))}
        </Select>
        <TextField
          id="standard-basic"
          label={`${EnumSearchType[searchType]} name`}
          value={searchResultsState.searchText}
          onChange={(e) => dispatch(changeSearchText(e.target.value))}
          inputProps={{
            "data-testid": "search-term",
          }}
        />
        <Button
          variant="contained"
          color="primary"
          onClick={() => {
            offset.current = 0;
            search(true);
          }}
          disabled={searchResultsState.searchText.trim().length === 0}
          data-testid="search-button"
        >
          Primary
        </Button>
      </FormControl>
    </div>
  );
};

export default SearchForm;
