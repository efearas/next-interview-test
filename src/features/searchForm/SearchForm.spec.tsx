import React from "react";
import { render, fireEvent, screen } from "@testing-library/react";
import SearchForm from "./SearchForm";
import { Provider } from "react-redux";
import { store } from "../../app/store";

describe("SearchForm", () => {
  it("should handle empty search term textbox", async () => {
    render(
      <Provider store={store}>
        <SearchForm />
      </Provider>
    );
    const inputSearchTerm = screen.getByTestId("search-term");
    const buttonSearch = screen.getByTestId("search-button");
    expect(buttonSearch).toHaveAttribute("disabled");
    fireEvent.change(inputSearchTerm, {
      target: { value: "Micheal Jackson" },
    });
    expect(buttonSearch).not.toHaveAttribute("disabled");
  });

  it("snapshot test", () => {
    const { container } = render(
      <Provider store={store}>
        <SearchForm />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });
});
