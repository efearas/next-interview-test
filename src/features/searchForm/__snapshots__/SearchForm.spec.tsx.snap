// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchForm snapshot test 1`] = `
<div>
  <div
    class="makeStyles-wrapper-4"
  >
    <div
      class="MuiFormControl-root makeStyles-formControl-3"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
        data-shrink="true"
        id="select-search-type"
      >
        Select search type
      </label>
      <div
        class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
      >
        <div
          aria-haspopup="listbox"
          aria-labelledby="select-search-type select-search-type"
          class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
          id="select-search-type"
          role="button"
          tabindex="0"
        >
          album
        </div>
        <input
          aria-hidden="true"
          class="MuiSelect-nativeInput"
          tabindex="-1"
          value="1"
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSelect-icon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
      <div
        class="MuiFormControl-root MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
          data-shrink="true"
          for="standard-basic"
          id="standard-basic-label"
        >
          artist name
        </label>
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input"
            data-testid="search-term"
            id="standard-basic"
            type="text"
            value="Micheal Jackson"
          />
        </div>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
        data-testid="search-button"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Primary
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;
